rust/chains/abacus-ethereum/src/multisig_ism.rs

```diff
@@ -0,0 +1,237 @@
+#![allow(clippy::enum_variant_names)]
+#![allow(missing_docs)]
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::{Duration, SystemTime};
+
+use abacus_core::accumulator::merkle::Proof;
+use async_trait::async_trait;
+use ethers::abi::Token;
+use ethers::providers::Middleware;
+use ethers::types::{Selector, H160, H256, U256};
+use eyre::Result;
+use std::collections::hash_map::Entry::Occupied;
+use std::hash::Hash;
+use tokio::sync::Mutex;
+
+use abacus_core::{
+    AbacusAbi, AbacusContract, ChainCommunicationError, ContractLocator, MultisigIsm,
+    MultisigSignedCheckpoint,
+};
+
+use crate::contracts::multisig_ism::{MultisigIsm as EthereumMultisigIsmInternal, MULTISIGISM_ABI};
+use crate::trait_builder::MakeableWithProvider;
+
+#[derive(Debug)]
+struct Timestamped<Value> {
+    t: SystemTime,
+    value: Value,
+}
+
+impl<Value> Timestamped<Value> {
+    fn new(value: Value) -> Timestamped<Value> {
+        Timestamped {
+            t: SystemTime::now(),
+            value,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ExpiringCache<Key, Value>
+where
+    Key: Eq + Hash,
+{
+    expiry: Duration,                        // cache endurance
+    cache: HashMap<Key, Timestamped<Value>>, // hashmap containing references to cached items
+}
+
+impl<Key, Value> ExpiringCache<Key, Value>
+where
+    Key: Copy + Eq + Hash + tracing::Value,
+    Value: Clone,
+{
+    pub fn new(expiry: Duration) -> ExpiringCache<Key, Value> {
+        ExpiringCache {
+            expiry,
+            cache: HashMap::new(),
+        }
+    }
+
+    pub fn put(&mut self, key: Key, value: Value) {
+        self.cache.insert(key, Timestamped::new(value));
+    }
+
+    pub fn get(&mut self, key: Key) -> Option<&Value> {
+        if let Occupied(entry) = self.cache.entry(key) {
+            if SystemTime::now()
+                .duration_since(entry.get().t)
+                .expect("Clock may have gone backwards")
+                > self.expiry
+            {
+                None
+            } else {
+                Some(&self.cache[&key].value)
```

Comment:
could you get the value from `entry` instead?

Reply:
Probably but I couldn't figure it out, I get the following when I try to return `Some(&entry.get().value)`

> cannot return value referencing local variable `entry`
returns a value referencing data owned by the current functionrustc[E0515](https://doc.rust-lang.org/error-index.html#E0515)

Per [1] it looks like I can do `Some(&entry.into_mut().value())` but I'm not sure I fully understand if this is better

[1] https://stackoverflow.com/questions/60129097/entryoccupied-get-returns-a-value-referencing-data-owned-by-the-current-func

Reply:
Using .get makes this all easier